dace/transformation/dataflow/map_unroll.py

```diff
@@ -0,0 +1,178 @@
+# Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
+
+from dace import dtypes, registry, SDFG
+from dace.sdfg import nodes, utils as sdutil
+from dace.sdfg.state import StateSubgraphView
+from dace.symbolic import evaluate
+from dace.transformation import transformation
+from dace.properties import make_properties
+import copy
+import itertools
+
+
+@registry.autoregister_params(singlestate=True)
+@make_properties
+class MapUnroll(transformation.Transformation):
+    """
+    Unrolls a map with constant ranges in the top-level scope of an SDFG by
+    replicating its subgraph for each iteration. If there are local data
+    containers only used in this map, they will also be replicated, as will
+    nested SDFGs found within.
+
+    This transformation can be useful for forming weakly connected components
+    that will be inferred as processing elements in an FPGA kernel.
+    """
+
+    _map_entry = nodes.MapEntry(nodes.Map("", [], []))
+
+    @staticmethod
+    def expressions():
+        return [sdutil.node_path_graph(MapUnroll._map_entry)]
+
+    @staticmethod
+    def can_be_applied(graph, candidate, expr_index, sdfg, strict=False):
+        map_entry = graph.nodes()[candidate[MapUnroll._map_entry]]
+        # Must be top-level map
+        if graph.scope_dict()[map_entry] is not None:
+            return False
+        # All map ranges must be constant
+        try:
+            for begin, end, step in map_entry.map.range:
+                evaluate(begin, sdfg.constants)
+                evaluate(end, sdfg.constants)
+                evaluate(step, sdfg.constants)
+        except TypeError:
+            return False
+        return True
+
+    @staticmethod
+    def match_to_str(graph, candidate):
+        map_entry = graph.nodes()[candidate[MapUnroll._map_entry]]
+        return map_entry.map.label + ': ' + str(map_entry.map.params)
+
+    def apply(self, sdfg):
+
+        from dace.transformation.dataflow import TrivialMapElimination
+
+        state = sdfg.nodes()[self.state_id]
+        map_entry = state.nodes()[self.subgraph[MapUnroll._map_entry]]
+        map_exit = state.exit_node(map_entry)
+
+        # Collect all nodes in this weakly connected component
```

Comment:
If I understand correctly, this is extracting the SDFGState's connected component that the Map resides in. Might me worth to make it a helper method (not necessarily in this PR, just mark it for later). Networkx has a method but only for undirected graphs